test/redis/interval_cache/getter_test.exs

```diff
@@ -0,0 +1,105 @@
+    assert_time Enum.at(hits, 3).time, "2017-03-22T02:00:00Z"
+    assert Enum.at(hits, 0).count == 11
+    assert Enum.at(hits, 1).count == 22
+    assert Enum.at(hits, 2).count == 33
+    assert Enum.at(hits, 3).count == 44
+  end
+
+  test "misses a partial range", %{from: from, to: to, rollup: rollup, keys: keys} do
+    Conn.hset(Enum.at(keys, 0), "key1", 11)
+    Conn.hset(Enum.at(keys, 1), "key1", 22)
+    Conn.hset(Enum.at(keys, 3), "key1", 44)
+    {hits, new_from} = Getter.get(@prefix, "key1", from, to, rollup)
+
+    assert length(hits) == 2
+    assert_time new_from, "2017-03-22T01:45:00Z"
+    assert_time Enum.at(hits, 0).time, "2017-03-22T01:15:00Z"
+    assert_time Enum.at(hits, 1).time, "2017-03-22T01:30:00Z"
+    assert Enum.at(hits, 0).count == 11
+    assert Enum.at(hits, 1).count == 22
+  end
+
+  test "misses the entire range", %{from: from, to: to, rollup: rollup, keys: keys} do
+    Conn.hset(Enum.at(keys, 1), "key1", 22)
+    Conn.hset(Enum.at(keys, 2), "key1", 33)
+    Conn.hset(Enum.at(keys, 3), "key1", 44)
+    {hits, new_from} = Getter.get(@prefix, "key1", from, to, rollup)
+    assert hits == []
+    assert_time new_from, "2017-03-22T01:15:00Z"
+  end
+
+  test "interprets hash-field misses as 0", %{from: from, to: to, rollup: rollup, keys: keys} do
+    Conn.hset(Enum.at(keys, 0), "key1", 11)
+    Conn.hset(Enum.at(keys, 1), "key2", 99)
+    Conn.hset(Enum.at(keys, 3), "key1", 44)
+    {hits, new_from} = Getter.get(@prefix, "key1", from, to, rollup)
+
+    assert length(hits) == 2
+    assert_time new_from, "2017-03-22T01:45:00Z"
+    assert_time Enum.at(hits, 0).time, "2017-03-22T01:15:00Z"
+    assert_time Enum.at(hits, 1).time, "2017-03-22T01:30:00Z"
+    assert Enum.at(hits, 0).count == 11
+    assert Enum.at(hits, 1).count == 0
+  end
+
+  test "hits all 0s", %{from: from, to: to, rollup: rollup, keys: keys} do
+    Conn.hset(Enum.at(keys, 0), "key2", 99)
+    Conn.hset(Enum.at(keys, 1), "key2", 99)
+    Conn.hset(Enum.at(keys, 2), "key2", 99)
+    Conn.hset(Enum.at(keys, 3), "key2", 99)
+    {hits, new_from} = Getter.get(@prefix, "key1", from, to, rollup)
+    assert length(hits) == 4
+    assert new_from == nil
+  end
+
+  test "assumes 0 on misses within a few seconds of now", %{from: from, to: to, rollup: rollup, keys: keys} do
+    Conn.hset(Enum.at(keys, 0), "key1", 11)
+    Conn.hset(Enum.at(keys, 1), "key1", 22)
+    Conn.hset(Enum.at(keys, 2), "key1", 33)
+    hits_at_time = fn(time) ->
+      with_mock Timex, [:passthrough], [now: fn() -> get_dtim(time) end] do
+        {hits, _new_from} = Getter.get(@prefix, "key1", from, to, rollup)
+        {length(hits), List.last(hits).count}
+      end
+    end
+    assert hits_at_time.("2017-03-22T03:00:00Z") == {3, 33}
+    assert hits_at_time.("2017-03-22T02:00:00Z") == {4, 0}
+    assert hits_at_time.("2017-03-22T02:00:29Z") == {4, 0}
+    assert hits_at_time.("2017-03-22T02:00:31Z") == {3, 33}
+    assert hits_at_time.("2017-03-22T01:00:00Z") == {4, 0}
```

Comment:
not quite sure what's going on in last one (line 103)? is it assuming that since the later 3 intervals have values, that that one didn't get set, and returns 0 for that interval?

Reply:
Yeah, so this is testing that we won't ask BQ for data about `2:00 - 2:15` until a few seconds after that interval starts.  There won't be any data there anyways, until kinesis/lambda have processed it.  (I have it set to 30 seconds right now, so it waits until `2:00:30` to ask).

The last line is an edge case test ... if it's `1:00` now, and you're asking for data an hour in the future ... I'm just going to assume it's 0.  Because how could BQ know the future?